examples/queue.rs

```diff
@@ -0,0 +1,91 @@
+/// FreeRTOS queue example
+/// The core concepts covered in this example include:
+/// - Creating and using queues to exchange data between an ISR(Interrupt Service Routine) and a thread (both directions).
+/// - Efficiently waiting for messages in the thread using queues.
+///
+/// This is demonstrated with a periodic timer ISR and the rust main thread, but it can be used in any thread or ISR context.
+use esp_idf_hal::delay::BLOCK;
+use esp_idf_hal::peripherals::Peripherals;
+use esp_idf_hal::task;
+use esp_idf_hal::task::queue::Queue;
+use esp_idf_hal::timer::{TimerConfig, TimerDriver};
+use esp_idf_sys::{uxQueueMessagesWaiting, EspError};
+
+fn main() -> Result<(), EspError> {
+    // It is necessary to call this function once. Otherwise some patches to the runtime
+    // implemented by esp-idf-sys might not link properly. See https://github.com/esp-rs/esp-idf-template/issues/71
+    esp_idf_hal::sys::link_patches();
+
+    let per = Peripherals::take()?;
+
+    let mut timer = TimerDriver::new(per.timer00, &TimerConfig::new().auto_reload(true))?;
+
+    // create two queue's that can hold 20/100 elements with the element size of an u8
+    let queue_send = Queue::<u8>::new(20);
+    let queue_recv = Queue::<u8>::new(100);
+
+    // SAFETY: as long as queue_send/recv live we can use isr_receive & isr_send.
+    let isr_recv: Queue<u8> = unsafe { Queue::new_borrowed(queue_send.as_raw()) };
+    let isr_send: Queue<u8> = unsafe { Queue::new_borrowed(queue_recv.as_raw()) };
+
+    // Every half a second
+    timer.set_alarm(timer.tick_hz() / 2)?;
+
+    // SAFTEY: make sure the timer is droped and thouse end the subscription before we drop queue_send/recv.
+    // E,g don't send stuff/recv stuff from a borrowd handle after that point. When the timer object is droped it will auto unsubscribe.
+    //
+    // Safe since we never end main and thouse never unsubscribe nor drop queue_send/queue_recv
+    unsafe {
+        timer.subscribe(move || {
```

Comment:
Since `&` is copy (unlike `&mut`) you can just push the references into `subscribe`. 
* If indeed `subscribe` is called, then `&` need to be transmuted to `&'static ` OR cast to `*const Queue<u8>` (same outcome)
* If you use `subscribe_nonstatic` here, none of the above ^^^ would be necessary, as `*_nonstatic` can use nonstatic borrows
* Finally, if you just `Arc` the queues (or even better, put then in a `'static` context with `static_cell`) also none of the above would be necessary and you can keep the usage of `subscribe` as opposed to switching to `subscribe_nonstatic`.